eva-accession-import/src/main/java/uk/ac/ebi/eva/accession/dbsnp/io/DbsnpVariantsWriter.java

```diff
@@ -83,4 +172,80 @@ private void writeClusteredVariantsDeclustered(List<DbsnpClusteredVariantEntity>
+                        List<OPERATION_ENTITY> merges = buildMergeOperations(variants,
+                                                                             hash,
+                                                                             mergedInto);
+
+                        operations.addAll(merges);
+                    });
+            return operations;
+        }
+
+        private Stream<String> extractUniqueHashes(BulkOperationException exception) {
+            return exception.getErrors()
+                            .stream()
+                            .filter(this::isDuplicateKeyError)
+                            .map(error -> {
+                                Matcher matcher = DUPLICATE_KEY_PATTERN.matcher(error.getMessage());
+
+                                if (!matcher.find()) {
+                                    throw new IllegalStateException(
+                                            "A duplicate key exception was caught, but the message couldn't be parsed correctly",
+                                            exception);
+                                } else {
+                                    String hash = matcher.group(DUPLICATE_KEY_ERROR_MESSAGE_GROUP_NAME);
+                                    return hash;
+                                }
+                            })
+                            .distinct();
+        }
+
+        private boolean isDuplicateKeyError(BulkWriteError error) {
+            ErrorCategory errorCategory = ErrorCategory.fromErrorCode(error.getCode());
+            return errorCategory.equals(ErrorCategory.DUPLICATE_KEY);
+        }
+
+        private List<OPERATION_ENTITY> buildMergeOperations(List<ENTITY> variants, String hash,
+                                                            ENTITY mergedInto) {
+            return removeDuplicatesWithSameHashAndAccession(variants.stream())
+                    .stream()
+                    .filter(v -> v.getHashedMessage().equals(hash)
+                            && !v.getAccession().equals(mergedInto.getAccession())
+                            && !isAlreadyMerged(v.getAccession()))
+                    .map(origin -> mergeOperationFactory.apply(origin, mergedInto))
+                    .collect(Collectors.toList());
+        }
+
+        private <T extends AccessionedDocument> Collection<T> removeDuplicatesWithSameHashAndAccession(
+                Stream<T> accessionedEntities) {
+            return accessionedEntities.collect(Collectors.toMap(v -> v.getHashedMessage() + v.getAccession().toString(),
+                                                                v -> v,
+                                                                (a, b) -> a))
+                                      .values();
+        }
+
+        private boolean isAlreadyMerged(Long accession) {
+            List<OPERATION_ENTITY> merges = operationRepository.findAllByAccession(accession);
```

Comment:
The type should be checked before assuming all of them are merges. Intuitively, there could be updates due to declustering.